clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -705,8 +705,18 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
       .Default(InvokeUnknown);
 }
 
+static const CXXRecordDecl *getKernelObjectType(QualType Ty) {
+  return Ty->getAsCXXRecordDecl();
+}
+
 static const CXXRecordDecl *getKernelObjectType(FunctionDecl *Caller) {
-  return (*Caller->param_begin())->getType()->getAsCXXRecordDecl();
+  assert(Caller->getNumParams() > 0 && "Insufficient kernel parameters");
+  return getKernelObjectType(Caller->getParamDecl(0)->getType());
+}
+
+static const CXXRecordDecl *getKernelObjectType(ArrayRef<const Expr *> Args) {
+  assert(Args.size() > 0 && "Insufficient kernel arguments");
+  return getKernelObjectType(Args[0]->getType());
 }
```

Comment:
@erichkeane , what is the rationale for splitting this function this way?  The other two are never called directly, right?

Reply:
Thew new patch uses the args to get the kernel object (the 3rd overload), which calls the 1st.  ConstructOpenCLKernel calls the FunctionDecl one.

The reason for splitting an otherwise simple looking functions was that the QualType case gets more complicated in another patch, where it also has to remove const-referenceness in some versions.

Initially, the FunctionDecl of the kernel wasn't passed to the CheckSYCLKernelCall, as it can operate on the Args themselves (and actually get a BETTER diagnostic location as a result).

Presumably, I could simplify this back down to making it unchanged, though it would make:

    const CXXRecordDecl *KernelObj = getKernelObjectType(Args);
    if (!KernelObj) {
      Diag(Args[0]->getExprLoc(), diag::err_sycl_kernel_not_function_object);

(see the Diag uses Args[0] and the 'get' uses 'Args') a touch more awkward.

That said, it is already a touch awkward, so I'll revert this part of the change.

Reply:
Thanks Erich, I missed the call from when it is constructed.  But I like the new version.